### PR TITLE
vsphere deploymentzone filter

### PR DESCRIPTION
**Pull request by sadysnaat**
**What this PR does / why we need it**:
adds support to filter deploymentzones by vspherecluster annotation 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/issues/1325

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note

```

### CODE REVIEW

File: controllers/vspherecluster_reconciler.go [hunk 1]
```diff
@@ -351,6 +353,23 @@ func (r clusterReconciler) reconcileDeploymentZones(ctx *context.ClusterContext)
 	return true, nil
 }
 
+func shouldIncludeZone(zone infrav1.VSphereDeploymentZone, cluster *infrav1.VSphereCluster) bool {
+	// expects cluster to be annotated
+	// vsphere.infrastructure.cluster.x-k8s.io/adopt-deploymentzone: "dz1, dz2"
+	// if there are there deploymentzones available
```
Comment (gab-satchi) [inline controllers/vspherecluster_reconciler.go:359]: small typo: should be `if there are three...`

File: controllers/vspherecluster_reconciler.go [hunk 2]
```diff
@@ -351,6 +353,23 @@ func (r clusterReconciler) reconcileDeploymentZones(ctx *context.ClusterContext)
...
+	// dz1, dz2 and dz3
+	// this annotation will allow cluster to include dz1 and dz2
+	deploymentZones, annotationExists := cluster.GetAnnotations()["vsphere.infrastructure.cluster.x-k8s.io/adopt-deploymentzone"]
```
Comment (gab-satchi) [inline controllers/vspherecluster_reconciler.go:362]: I'd suggest to move that annotation key into a constant and maybe add some unit tests under `vsphere_cluster_reconciler_test.go`

**Review by gab-satchi — COMMENTED** [review]
changes look good. I'll wait on @srm09 to comment on why we omitted this customization ability on the original design. I'd also suggest adding something to our docs so users can know about the annotation. I believe there's a doc issue already for deploymentZones so can be punted till then.

File: controllers/vspherecluster_reconciler.go (hunk 1 above)
Comment (sadysnaat) [reply controllers/vspherecluster_reconciler.go]: Ack

File: controllers/vspherecluster_reconciler.go (hunk 2 above)
Comment (sadysnaat) [reply controllers/vspherecluster_reconciler.go]: Ack

**Comment by srm09** [comment]
@sadysnaat Reviving this one and hoping to get it merged soon enough. 
Could you rebase this PR with the latest on the `main` branch, plus can you also include the changes to the original Failure Domain proposal doc, so that it is up-to-date with the latest additions.
I will review the PR once again after this is done, thanks for being patient.

**Comment by srm09** [comment]
@sadysnaat Whenever you have some time, could you please take a look at this one?

**Comment by sadysnaat** [comment]
> @sadysnaat Whenever you have some time, could you please take a look at this one?

sure @srm09 will pick this up

**Comment by srm09** [comment]
We are rethinking the support for Multi AZ support in CAPV, and I will close this for now. I will keep the community up to date around the latest improvements around multi AZ
/close

**Pull request closed without merging**